[AsyncFlush 3/3] Use asynchronous flush

nbiscaro opened the pull request:
<!-- .github/PULL_REQUEST_TEMPLATE.md -->

## Summary

Final PR for implementing asynchronous flush events for both streaming and non-streaming transactions.

Previous PRs: 
Use globally unique batch ids: https://github.com/Mooncake-Labs/moonlink/pull/1031
Add streaming and non-streaming state to same snapshot: https://github.com/Mooncake-Labs/moonlink/pull/1100

In this PR, we change all table flush to be asynchronous. We spawn the flush as background task and send `TableEvent::FlushResult` on complete, at which point we apply the results of the flush. Given we may have multiple concurrent flushes at once we track a queue of pending flush LSNs. Before iceberg snapshot we ensure there is no pending LSN less than the LSN we want to persist at. 

## Related Issues

Closes https://github.com/Mooncake-Labs/moonlink/issues/234

## Changes

- 
- 
- 

## Checklist

- [ ] Code builds correctly
- [ ] Tests have been added or updated
- [ ] Documentation updated if necessary
- [ ] I have reviewed my own changes

dentiny commented:
FYI: I did a few fixes yesterday and today, suggest to rebase upon main.

nbiscaro commented on src/moonlink/src/table_notify.rs:
```diff
@@ -170,6 +171,12 @@ pub enum TableEvent {
     PeriodicalPersistTruncateWalResult {
         result: Result<WalPersistAndTruncateResult>,
     },
+    FlushResult {
```
Added its own debug formatter that only prints the batch ids in the disk slice

dentiny commented:
```sh
Test test_chaos_on_local_fs_with_index_merge is with random seed 1754428183492768295
Test test_chaos_on_local_fs_with_data_compaction is with random seed 1754428183492836537
Test test_chaos_on_local_fs_with_no_background_maintenance is with random seed 1754428183493092372

thread 'tokio-runtime-worker' panicked at src/moonlink/src/storage/mooncake_table/test_utils.rs:175:5:
assertion `left == right` failed
  left: [3, 7, 10, 13, 16, 18, 19, 22, 23, 26, 27, 28, 30, 32, 33, 34, 35, 36, 37, 38, 39, 40, 43, 44, 45, 46, 48, 49, 50]
 right: [3, 7, 10, 13, 16, 18, 19, 22, 23, 26, 27, 28, 30, 32, 33, 34, 35, 36, 37, 38, 39, 40, 43, 44, 45, 46, 49, 50]
```
I used the random seed to check in the main branch for a few, cannot repro, likely new issue introduce in this PR, but not existing ones.

dentiny commented:
```sh
thread 'tokio-runtime-worker' panicked at src/moonlink/src/storage/mooncake_table/snapshot.rs:893:21:
assertion failed: res
stack backtrace:
   0: __rustc::rust_begin_unwind
```
which comes from 
https://github.com/Mooncake-Labs/moonlink/blob/8bcd9ce5af9163a0a74e7c753099635676863070/src/moonlink/src/storage/mooncake_table/snapshot.rs#L887-L893, seems still an issue?

dentiny commented on src/moonlink/src/table_handler/tests.rs:
```diff
@@ -2118,3 +2103,35 @@ fn test_deletion_vector_capacity_exceeded_minimal() {
 }
+
+#[tokio::test]
+async fn test_stream_delete_from_stream_memslice_row_debug() {
```
What does it mean by `_debug`? Should we clean it up?

dentiny commented on src/moonlink/src/table_handler/tests.rs:
```diff
@@ -2118,3 +2103,35 @@ fn test_deletion_vector_capacity_exceeded_minimal() {
+
+    // Phase 4: Commit the streaming transaction
+    env.stream_commit(stream_commit_lsn, xact_id).await;
+    println!("Debug: Verifying final state post-commit (should be empty)");
```
Should we clean it up? I mean it says "debug".

nbiscaro replied:
removed

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -687,6 +691,11 @@ impl MooncakeTable {
     }
 
+    #[cfg(test)]
+    pub(crate) fn set_iceberg_snapshot_lsn(&mut self, lsn: u64) {
```
Don't we already have a `set_iceberg_snapshot_res`?
And we could always set iceberg snapshot LSN at construction?

nbiscaro replied:
`set_iceberg_snapshot_res` takes many complex parameters that are not relevant for this test. I think directly setting the lsn is clean and accomplishes only what we are trying to do. 

If you are concerned about the `#[cfg(test)]` being in with our source code, I have moved it to its own `impl` in the `table_creation_test_utils` file. 

We should actually do the same with any other test specific code we have behind this flag. Then we don't have to implement hacky workarounds when testing and can access the data we need without cluttering or complicating actual implementation. https://github.com/Mooncake-Labs/moonlink/issues/1261

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+        if let Some(batch) = new_batch {
+            self.next_snapshot_task
+                .new_record_batches
+                .push((batch.0, batch.1, None));
```
let me try again, do you mind we using struct instead of tuples? Struct provides field naming, which is much more readable.

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+        }
+
+        tokio::task::spawn(async move {
+            let mut disk_slice_clone = DiskSliceWriter::new(
```
Curious why name it `clone`?

nbiscaro replied:
Ah, this was before disk slice was cloneable. we can just directly clone now.

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+            match flush_result {
+                Ok(()) => {
+                    table_notify_tx
+                        .send(TableEvent::FlushResult {
```
(not sth we need to address now) If a disk write fails, do we leave orphaned disk files?

nbiscaro replied:
Probably. We also leave orphaned files on stream abort. Followup created https://github.com/Mooncake-Labs/moonlink/issues/1260

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+                        .unwrap();
+                }
+                Err(e) => {
+                    table_notify_tx
```
You need to coordinate with drop table, otherwise send would fail.

dentiny replied:
https://github.com/Mooncake-Labs/moonlink/blob/dff3672daad96bca992de37315fd19f529b844ad/src/moonlink/src/table_handler/table_handler_state.rs#L327-L341

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
         Ok(())
     }
 
+    fn set_next_flush_lsn(&mut self, lsn: u64) {
```
Need documentation or rename, not easy to tell.

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+        let lsn = disk_slice
+            .lsn()
+            .expect("LSN should never be none for non streaming flush");
+        self.pending_flush_lsns.remove(&lsn);
```
Please check return value.

dentiny commented on src/moonlink/src/storage/mooncake_table/transaction_stream.rs:
```diff
@@ -343,36 +342,48 @@ impl MooncakeTable {
     /// Removes in memory indices and record batches from the stream state.
     pub fn apply_stream_flush_result(&mut self, xact_id: u32, mut disk_slice: DiskSliceWriter) {
+        if let Some(lsn) = disk_slice.lsn() {
+            self.pending_flush_lsns.remove(&lsn);
```
Please check return value.

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -446,6 +446,9 @@ pub struct MooncakeTable {
     wal_manager: WalManager,
+
+    /// LSN of pending flushes.
+    pub pending_flush_lsns: BTreeSet<u64>,
```
`pending` means not yet decided: https://www.merriam-webster.com/dictionary/pending
I'm not sure if it means (1) ongoing, (2) in flush queue, if we have one; (3) in-memory, will be flushed.
```suggestion
    pub ongoing_flush_lsns: BTreeSet<u64>,
```

nbiscaro replied:
changed

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -843,14 +940,20 @@ impl MooncakeTable {
         self.next_snapshot_task = SnapshotTask::new(self.metadata.config.clone());
+        // Carry forward the commit baseline
+        // This is important if we have a pending flush that will be added to the next snapshot task.
+        self.next_snapshot_task.new_commit_lsn = next_snapshot_task.new_commit_lsn;
```
- You need to rename the variable, it's not only **new** LSNs any more
- Could you please elaborate what if we don't carry over? 

TBH
> This is important if we have a pending flush that will be added to the next snapshot task.

Tells me nothing

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -843,14 +940,20 @@ impl MooncakeTable {
 
         let cur_snapshot = self.snapshot.clone();
+
+        let has_pending_flushes = !self.pending_flush_lsns.is_empty();
```
```suggestion
        let has_ongoing_flushes = !self.ongoing_flush_lsns.is_empty();
```

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
+        }
+    }
+
+    pub fn get_min_pending_flush_lsn(&self) -> u64 {
```
nit: doc on fallback behavior

dentiny commented on src/moonlink/src/storage/mooncake_table/disk_slice.rs:
```diff
@@ -51,6 +53,16 @@ pub struct DiskSliceWriter {
+        f.debug_struct("DiskSliceWriter")
+            .field("batch_ids", &batch_ids)
+            .field("lsn", &self.writer_lsn)
+            .finish()
```
I strongly suggest to print out the number of data files and index blocks, they're super useful in debugging to tell whether we should trigger compaction and index merge.

nbiscaro replied:
added

dentiny commented on src/moonlink/src/storage/mooncake_table/snapshot.rs:
```diff
@@ -562,7 +563,10 @@ impl SnapshotTableState {
-        if !opt.skip_iceberg_snapshot && (force_empty_iceberg_payload || flush_by_table_write) {
+        if !opt.skip_iceberg_snapshot
+            && (force_empty_iceberg_payload || flush_by_table_write)
+            && !has_pending_flushes
```
I don't think it's good (it's correct behavior though), because it means we cannot make iceberg snapshots whenever we have **ANY** background flushes.

nbiscaro replied:
fixed to use correct condition of `flush_lsn < min_ongoing_flush_lsn`

dentiny commented on src/moonlink/src/storage/mooncake_table/snapshot.rs:
```diff
@@ -701,21 +705,34 @@ impl SnapshotTableState {
+                    slice.remap_deletion_if_needed(deletion)
+                {
+                    // Find the disk file entry and apply the deletion
+                    for (file_ref, disk_entry) in self.current_snapshot.disk_files.iter_mut() {
```
nit (feel free to ignore): one thing I plan to do is get the dependency graph for updating snapshot -- I made mistakes to reorder a few steps.
For example, integrating disk slice function depends on all new data files have been reflected to disk file map; would appreciate if you could mark the precondition.

nbiscaro replied:
added

dentiny commented on src/moonlink/src/storage/mooncake_table/transaction_stream.rs:
```diff
@@ -512,9 +512,21 @@ impl MooncakeTable {
         Ok(())
     }
+
+    /// Commit a transaction stream.
```
```suggestion
    /// Commit a transaction stream.
    /// - Flushes any remaining rows from stream mem slice
    /// - Adds all in mem batches and indices to next snapshot task
    /// - Updates deletion records
    /// - Enqueues `TransactionStreamOutput::Commit` for snapshot task
```

nbiscaro commented:
> Looks like an issue:
> 
> ```shell
> thread 'tokio-runtime-worker' panicked at src/moonlink/src/storage/mooncake_table/snapshot.rs:394:17:
> assertion failed: old_entry.puffin_deletion_blob.is_none()
> stack backtrace:
>    0: __rustc::rust_begin_unwind
> ```
> 
> https://github.com/Mooncake-Labs/moonlink/actions/runs/16791181367/job/47552933256?pr=1204

For a minute I thought we were forgetting to remap local deletions in the case where we are applying flush post-commit, so i applied them there and pushed the change. I now realize we remap these in `integrate_disk_slices` so this was incorrect. Removed that change and seems fine.

nbiscaro commented:
uhhh i don't know why github closed that...

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -283,7 +290,7 @@ impl SnapshotTask {
     pub fn should_create_snapshot(&self) -> bool {
         // If mooncake has new transaction commits.
-        self.new_commit_lsn > 0
+        self.commit_lsn_baseline > 0
```
issue here: we will create new mooncake snapshot every 500ms

nbiscaro replied:
added `previous_commit_lsn_baseline` to compare against and track changes.

dentiny commented on src/moonlink/src/storage/mooncake_table/transaction_stream.rs:
```diff
@@ -424,27 +423,28 @@ impl MooncakeTable {
+        let mut stream_state = self
+            .transaction_stream_states
+            .remove(&xact_id)
+            .expect("Stream state not found for xact_id: {xact_id}");
```
suggest to attach LSN as well, so ourselves could differentiate stream flush and stream commit

dentiny commented on src/moonlink/src/table_handler.rs:
```diff
@@ -679,16 +697,24 @@ impl TableHandler {
+        let should_force_snapshot = TableHandlerState::should_force_snapshot_by_commit_lsn(
+            lsn,
+            min_pending_flush_lsn,
+            &table_handler_state.table_maintenance_process_status,
```
not sure why not using member function here?

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -724,7 +739,7 @@ impl MooncakeTable {
 
     pub fn commit(&mut self, lsn: u64) {
-        self.next_snapshot_task.new_commit_lsn = lsn;
+        self.next_snapshot_task.commit_lsn_baseline = lsn;
```
assert LSN never goes back

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +769,166 @@ impl MooncakeTable {
     }
 
+    // Sets the flush LSN for the next iceberg snapshot. Note that we can only set the flush LSN if it's greater than the current min pending flush LSN. Otherwise, LSNs will be persisted to iceberg in the wrong order.
+    fn set_next_flush_lsn(&mut self, lsn: u64) {
```
The naming is confusing, it's an "attempt", no guarantee to set LSN

nbiscaro replied:
changed to `try_set_next_flush_lsn`

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +769,166 @@ impl MooncakeTable {
         Ok(())
     }
 
+    // Sets the flush LSN for the next iceberg snapshot. Note that we can only set the flush LSN if it's greater than the current min pending flush LSN. Otherwise, LSNs will be persisted to iceberg in the wrong order.
```
```suggestion
    // Attempt to set the flush LSN for the next iceberg snapshot. Note that we can only set the flush LSN if it's greater than the current min pending flush LSN. Otherwise, LSNs will be persisted to iceberg in the wrong order.
```
or sth alike

nbiscaro replied:
It makes no such assumption. If we can set a new LSN we will, otherwise it remains None which is also valid. I changed the name to `try_set...` to make this more clear.

dentiny commented on src/moonlink/src/storage/mooncake_table/transaction_stream.rs:
```diff
@@ -343,36 +342,36 @@ impl MooncakeTable {
+            if disk_slice.writer_lsn.is_none() {
+                disk_slice.writer_lsn = Some(commit_lsn);
+            }
+
```
nit: it reads to me as long as pending (should rename to `ongoing` for consistency) count is 0, we remove from stream state, better to extract it out, maybe right after L360.

nbiscaro replied:
that is whats happening. If we commit && there are no ongoing flushes we remove the stream state.

dentiny reviewed (commented):
Second round...

dentiny commented on src/moonlink/src/storage/mooncake_table/disk_slice.rs:
```diff
@@ -247,20 +268,26 @@ impl DiskSliceWriter {
     }
 
-    pub fn remap_deletion_if_needed(&self, deletion: &mut ProcessedDeletionRecord) {
+    pub fn remap_deletion_if_needed(
```
nit: comment on return value (i.e., when to return None)

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -1112,13 +1216,14 @@ impl MooncakeTable {
             .write()
             .await
-            .update_snapshot(next_snapshot_task, opt)
+            .update_snapshot(next_snapshot_task, opt, min_pending_flush_lsn)
```
curious why don't we place min flush LSN inside of snapshot task?

nbiscaro replied:
changed

dentiny commented on src/moonlink/src/storage/mooncake_table/snapshot.rs:
```diff
@@ -525,8 +526,8 @@ impl SnapshotTableState {
 
-        if task.new_commit_lsn != 0 {
-            self.current_snapshot.snapshot_version = task.new_commit_lsn;
+        if task.commit_lsn_baseline != 0 {
```
nit: I feel like we could directly assign?
```rs
self.current_snapshot.snapshot_version = task.commit_lsn_baseline;
```

dentiny replied:
Also please add assertion so we're sure LSN never goes back

dentiny commented on src/moonlink/src/storage/mooncake_table/snapshot.rs:
```diff
@@ -611,15 +615,15 @@ impl SnapshotTableState {
         let (streaming_batches, mut non_streaming_batches): (Vec<_>, Vec<_>) = incoming
             .into_iter()
-            .partition(|(id, _, _)| *id < (1u64 << 63));
+            .partition(|batch| batch.batch_id < (1u64 << 63));
```
making it a constant? external contributors won't understand what's the logic here

nbiscaro replied:
made constant

dentiny commented on src/moonlink/src/storage/mooncake_table/snapshot.rs:
```diff
@@ -701,21 +706,35 @@ impl SnapshotTableState {
+                    < write_lsn.expect(
                         "Critical: LSN is None after it should have been updated by commit process",
-                    )
+                    ) - 1
```
I don't think there's anything functionally wrong here (attempt remap is possible doesn't hurt anyway)
But could you please help me understand on the write lsn - 1? Thank you!

nbiscaro replied:
Added comment 

```
            // remap deletions written *after* this slice’s LSN
            // We set to write_lsn - 1 to maintain consistency with deletion LSNs from the streaming case.
            // In the case a streaming transaction commits before flush, we assign disk_slice.writer_lsn = commit_lsn.
            // The corresponding deletions have lsn = commit_lsn - 1, so we need >= write_lsn - 1
            // to ensure these deletions get remapped from memory to disk locations.

dentiny commented on src/moonlink/src/storage/mooncake_table/transaction_stream.rs:
```diff
@@ -343,36 +342,36 @@ impl MooncakeTable {
     /// Handles commit and abort cleanup.
     /// Removes in memory indices and record batches from the stream state.
     pub fn apply_stream_flush_result(&mut self, xact_id: u32, mut disk_slice: DiskSliceWriter) {
+        if let Some(lsn) = disk_slice.lsn() {
```
(definitely not right now) we need to leave a TODO/issue to revisit on error handling, since we will have bad streaming state kept in memory forever -> OOM

dentiny commented on src/moonlink/src/table_handler/failure_tests.rs:
```diff
@@ -329,7 +355,8 @@ async fn test_force_data_compaction_with_failed_iceberg_persistence() {
     .await;
     env.commit(/*lsn=*/ 20).await;
-    env.flush_table(/*lsn=*/ 30).await;
+    // env.flush_table(/*lsn=*/ 30).await;
```
remove

dentiny commented on src/moonlink/src/storage/mooncake_table/tests.rs:
```diff
@@ -1142,38 +1219,1290 @@ async fn test_streaming_begin_flush_abort_end_flush_multiple() {
+}
+
+#[tokio::test]
+async fn test_stream_commit_with_pending_flush_deletion_remapping_bug() -> Result<()> {
```
why name the test case as `_bug`?

dentiny replied:
If you think it's a bug fix, you need to link the issue around (duckdb does that)

dentiny commented on src/moonlink/src/storage/mooncake_table/tests.rs:
```diff
@@ -1142,38 +1219,1290 @@ async fn test_streaming_begin_flush_abort_end_flush_multiple() {
+    // This should persist the deletion vectors as puffin blobs
+    create_mooncake_and_persist_for_test(&mut table, &mut event_completion_rx).await;
+
+    // Step 4: Add more data to trigger compaction threshold
```
FYI: you should be able to tune the compaction threshold

dentiny commented on src/moonlink/src/storage/mooncake_table/tests.rs:
```diff
@@ -1142,38 +1219,1290 @@ async fn test_streaming_begin_flush_abort_end_flush_multiple() {
+    });
+    assert!(created, "Mooncake snapshot should be created");
+
+    // This should succeed with the fix, but would fail without it
```
same suggestion here, if it's a bug fix, should link some reference
For example, duckdb puts issue fix tests under a certain directory: https://github.com/duckdb/duckdb/tree/main/test/issues/general

dentiny commented on src/moonlink/src/storage/mooncake_table/tests.rs:
```diff
@@ -1142,38 +1219,1290 @@ async fn test_streaming_begin_flush_abort_end_flush_multiple() {
+}
+
+#[tokio::test]
+async fn test_chaos_sequence_exact_repro_with_async_flush() -> Result<()> {
```
what does the test name mean lol

dentiny commented on src/moonlink/src/table_handler.rs:
```diff
@@ -526,6 +527,23 @@ impl TableHandler {
+                                        table.apply_flush_result(disk_slice);
+                                    }
+                                }
+                                Some(Err(e)) => {
```
let's unwrap it to keep the same behavior (prev sync flush unwraps directly)

My concern is:
- for now, for failed async flush, when an error happens we do nothing; no snapshot/mooncake table state update, no crash
- for chaos test it failed right away in the past, since event handler thread crashes; now it needs to wait for timeout on CI, and with no events printed out

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +769,166 @@ impl MooncakeTable {
-        if self.mem_slice.is_empty() {
+        let table_notify_tx = self.table_notify.as_ref().unwrap().clone();
+
+        if self.mem_slice.is_empty() || self.ongoing_flush_lsns.contains(&lsn) {
```
(just a question, not request to change or update) do we have unit test for no mem slice to flush?

dentiny replied:
I don't understand the second half, the only scenario I could think of where we have duplicate flush LSN is force snapshot, shouldn't we skip directly?

nbiscaro replied:
> (just a question, not request to change or update) do we have unit test for no mem slice to flush?

defs covered in existing tests (not sure if there is an explicit isolated test)

> I don't understand the second half, the only scenario I could think of where we have duplicate flush LSN is force snapshot, shouldn't we skip directly?

Sure can remove

dentiny commented on src/moonlink/src/storage/mooncake_table.rs:
```diff
@@ -754,84 +769,166 @@ impl MooncakeTable {
+
+        if self.mem_slice.is_empty() || self.ongoing_flush_lsns.contains(&lsn) {
+            self.set_next_flush_lsn(lsn);
+            tokio::task::spawn(async move {
```
I don't understand why do we need to send an empty flush result back to eventloop, isn't updating flush LSN already enough? (I mean, eventloop treats it as no-op)

dentiny replied:
That way, you don't need the `Option` wrapper

nbiscaro replied:
I think it's a cleaner mental model to notify the table that the flush process is complete regardless whether the flush has/does not have data or was successful/unsuccessful. It allows us to verify the flush process finished all the way through.

dentiny commented on src/moonlink/src/storage/mooncake_table/table_operation_test_utils.rs:
```diff
@@ -24,11 +26,124 @@ use crate::{ReadState, Result};
+    }
+}
+
+/// Flush mooncake, block wait its completion and reflect result to mooncake table.
```
The comment looks suspicious, I don't see the apply result part

dentiny commented on src/moonlink/src/storage/mooncake_table/table_operation_test_utils.rs:
```diff
@@ -24,11 +26,124 @@ use crate::{ReadState, Result};
+    }
+}
+
+/// Commit transaction stream, block wait its completion and reflect result to mooncake table.
```
Shouldn't we `apply_stream_flush_result` to sync result to table?

nbiscaro commented on src/moonlink/src/table_handler.rs:
```diff
@@ -376,7 +376,8 @@ impl TableHandler {
                             // Process iceberg snapshot and trigger iceberg snapshot if necessary.
-                            if table_handler_state.can_initiate_iceberg_snapshot() {
+                            let min_pending_flush_lsn = table.get_min_pending_flush_lsn();
+                            if TableHandlerState::can_initiate_iceberg_snapshot(lsn, min_pending_flush_lsn, table_handler_state.iceberg_snapshot_result_consumed, table_handler_state.iceberg_snapshot_ongoing) {
```
I want to directly test without initializing the entire TableHandlerState object

nbiscaro commented:
> 1. We inherit commit LSN, it seems we create mooncake snapshot every 500ms?
>    https://github.com/Mooncake-Labs/moonlink/pull/1204/files#r2261138006
> 2. Duplicate flush LSN handling read weird, I don't understand the necessity for an empty flush result; same for empty mem slice
> 3. Error handling behavior is different with before, which could block chaos test forever, not a good behavior.
>    3.1 I left a comment on failure related testing and chaos injection test infra, I **hope** (not a requirement) we could do the former in this PR, and leave chaos injection to next PR
> 4. remove completed lsn and set flush lsn needs better documentation/API, it's error prune without any assertion (actually if you take my suggestion item 2, you don't need two functions). Set flush LSN needs to be renamed, personal suggestion.
> 5. Not sure why you refactor table state handler's member function into class function, that makes code verbose (have to pass 4+ arguments); more importantly, we should keep coding style consistent (i.e., can_drop_table_now is still a member function)
> 6. naming consistency, should use "ongoing" instead of "pending" everywhere
> 7. I don't understand the deletion log partition LSN logic (https://github.com/Mooncake-Labs/moonlink/pull/1204/files#r2261249891), appreciate some explanation. Thank you!

1. Good catch, changed to use `previous_commit_baseline_lsn` to compare against and determine if the snapshot is dirty or not. 
2. removed
3. Changed to panic on error
3.1 Added test for disk write failure
4. added docs
5. So we can test that logic directly without needed to init entire table handler state
6. searched and replaced
7. added comment inline


Thank you for review :)

dentiny commented:
followup on https://github.com/Mooncake-Labs/moonlink/pull/1204/files#r2261670281
Don't see changes

dentiny commented on src/moonlink/src/table_notify.rs:
```diff
@@ -180,6 +181,12 @@ pub enum TableEvent {
         result: Result<WalPersistenceUpdateResult>,
     },
+    FlushResult {
+        // Transaction ID
```
You're putting it in the wrong place
